# Buổi 03 - Bài 02 - Bài Tập Tổng Hợp

- MSSV: 25410291
- Họ Tên: Đinh Xuân Sâm
- GDrive: [IE313-b03-Bai02.ipynb](https://colab.research.google.com/drive/1N-pNcXCixloVnUNra3I9uUEtx3ix6WB7?usp=sharing)


Yêu cầu:

Sử dụng bộ dữ liệu thô ở chương 1 (link trong slide). Sau đó, tiến hành xử lý theo quy trình sau:

1. Cột `price` => `dropna` => dòng.
2. Cột `horsepower` => `replace` => `mean`.
3. Cột `normalized-losses` => `replace` => `mode`.
4. Bước cuối cùng, `dropna` dòng cho toàn bộ dữ liệu.
5. Quan sát giá trị thực tế, sau đó chuyển kiểu phù hợp cho các cột còn lại.
6. Chọn ra các cột số, chuẩn hóa thành `zscore`.

Nộp file code dạng `ipynb`.

## 0. Nạp Dữ Liệu

In [1]:
# Đạt biến chứa url/tên và nếu dùng gdrive cho dataset
dataset_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"
dataset_file = "imports-85.csv"
dataset_file_gdrive = "/content/drive/MyDrive/uit/courses/IE313/06_Dataset/imports-85.csv"

In [2]:
# Tải dataset
import requests

response = requests.get(dataset_url)

with open(dataset_file, "wb") as file:
    file.write(response.content)

print("Download complete!")

Download complete!


In [3]:
# dataset headers
headers = [
"symboling",
"normalized-losses",
"make",
"fuel-type",
"aspiration",
"num-of-doors",
"body-style",
"drive-wheels",
"engine-location",
"wheel-base",
"length",
"width",
"height",
"curb-weight",
"engine-type",
"num-of-cylinders",
"engine-size",
"fuel-system",
"bore","stroke",
"compression-ratio",
"horsepower",
"peak-rpm",
"city-mpg",
"highway-mpg",
"price"]

In [4]:
# import panda
import pandas as pd

In [5]:
# đọc file dataset với headers chỉ định
df = pd.read_csv(dataset_file, names=headers)

In [6]:
# df quick look
df

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
0,3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,13495
1,3,?,alfa-romero,gas,std,two,convertible,rwd,front,88.6,...,130,mpfi,3.47,2.68,9.0,111,5000,21,27,16500
2,1,?,alfa-romero,gas,std,two,hatchback,rwd,front,94.5,...,152,mpfi,2.68,3.47,9.0,154,5000,19,26,16500
3,2,164,audi,gas,std,four,sedan,fwd,front,99.8,...,109,mpfi,3.19,3.40,10.0,102,5500,24,30,13950
4,2,164,audi,gas,std,four,sedan,4wd,front,99.4,...,136,mpfi,3.19,3.40,8.0,115,5500,18,22,17450
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
200,-1,95,volvo,gas,std,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,9.5,114,5400,23,28,16845
201,-1,95,volvo,gas,turbo,four,sedan,rwd,front,109.1,...,141,mpfi,3.78,3.15,8.7,160,5300,19,25,19045
202,-1,95,volvo,gas,std,four,sedan,rwd,front,109.1,...,173,mpfi,3.58,2.87,8.8,134,5500,18,23,21485
203,-1,95,volvo,diesel,turbo,four,sedan,rwd,front,109.1,...,145,idi,3.01,3.40,23.0,106,4800,26,27,22470


## 1. Cột `price` => `dropna` => dòng.

In [7]:
# Kích thước của df hiện tại
df.shape

(205, 26)

In [8]:
# Thông tin của df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

In [9]:
# Kích thước của price
df['price'].shape

(205,)

In [10]:
# import numpy
import numpy as np

In [11]:
# Thay thế giá trị khuyết tự nhiên thành Numpy N/A và kiểm tra khuyết
df.replace(to_replace='?', value=np.nan, inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

### Nhận xét

- `price` có 201 dòng `non-null`, có nghĩa có 4 dòng N/A cần được loại bỏ

In [12]:
# Thay thế giá trị khuyết tự nhiên thành Numpy N/A, inplace
df.replace(to_replace='?', value=np.nan, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       203 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

In [13]:
# Kích thước của price sau khi điền N/A
df['price'].shape

(205,)

In [14]:
# Kiểm thử trước khi dropna
df.dropna(axis=0, subset=['price'], inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non-null

In [15]:
# Thực hiện dropna và kiểm tra lại kích thước của price
df.dropna(axis=0, subset=['price'], inplace=True)
df['price'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: price
Non-Null Count  Dtype 
--------------  ----- 
201 non-null    object
dtypes: object(1)
memory usage: 3.1+ KB


In [16]:
# Kích thước của price sau khi dropna
df['price'].shape

(201,)

In [17]:
# Kiểm tra lại kích thước của df
df.shape

(201, 26)

### Kết Quả

- `price` đã được `dropna`.
- Số lượng dòng GIẢM từ 205 xuống 201 (tương ứng số dòng `non-null` đã đề cập trước đó).
- Số dòng của `df` GIẢM tương ứng vì dòng sẽ được xóa tương ứng.

## 2. Cột `horsepower` => `replace` => `mean`.

### Đánh Giá

- `horsepower` hiện đang có kiểu là `object`.
- Cần chuyển kiểu dữ liệu về kiểu số (`float`/`int`) để có thể tính `mean`.

In [18]:
# Lấy sample để đối chiếu kiểu dữ liệu
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
97,1,103,nissan,gas,std,four,wagon,fwd,front,94.5,...,97,2bbl,3.15,3.29,9.4,69,5200,31,37,7999


In [19]:
# Kiểu dữ liệu hiện tại
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 201 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          201 non-null    int64  
 1   normalized-losses  164 non-null    object 
 2   make               201 non-null    object 
 3   fuel-type          201 non-null    object 
 4   aspiration         201 non-null    object 
 5   num-of-doors       199 non-null    object 
 6   body-style         201 non-null    object 
 7   drive-wheels       201 non-null    object 
 8   engine-location    201 non-null    object 
 9   wheel-base         201 non-null    float64
 10  length             201 non-null    float64
 11  width              201 non-null    float64
 12  height             201 non-null    float64
 13  curb-weight        201 non-null    int64  
 14  engine-type        201 non-null    object 
 15  num-of-cylinders   201 non-null    object 
 16  engine-size        201 non-null

Các cột cần đổi kiểu dữ liệu:

```
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object -> float
 2   make               205 non-null    object
 3   fuel-type          205 non-null    object
 4   aspiration         205 non-null    object
 5   num-of-doors       205 non-null    object
 6   body-style         205 non-null    object
 7   drive-wheels       205 non-null    object
 8   engine-location    205 non-null    object
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object
 15  num-of-cylinders   205 non-null    object -> int
 16  engine-size        205 non-null    int64  
 17  fuel-system        205 non-null    object
 18  bore               205 non-null    object -> float
 19  stroke             205 non-null    object -> float
 20  compression-ratio  205 non-null    float64
 21  horsepower         205 non-null    object -> float
 22  peak-rpm           205 non-null    object -> int
 23  city-mpg           205 non-null    int64  
 24  highway-mpg        205 non-null    int64  
 25  price              205 non-null    object -> float
```

In [20]:
# Đổi kiểu dữ liệu để phản ánh đúng kiểu của giá trị
df['horsepower'] = df['horsepower'].astype('float')
df['horsepower'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
199 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [21]:
# Tính mean của horsepower
mean_horsepower = df['horsepower'].mean()
mean_horsepower

np.float64(103.39698492462311)

In [22]:
# Kiểm tra trước kết quả của phép thay thế NA với mean
df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=False).info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [23]:
# Thực hiện phép thay thế NA với mean
df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=True)
df['horsepower'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: horsepower
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


/tmp/ipykernel_23666/889122992.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['horsepower'].replace(to_replace=np.nan, value=mean_horsepower, inplace=True)


In [24]:
# Kiểm tra lại
df['horsepower']

,horsepower
0,111.0
1,111.0
2,154.0
3,102.0
4,115.0
...,...
200,114.0
201,160.0
202,134.0
203,106.0


### Kết Quả

- `horsepower` đã được đổi kiểu dữ liệu từ `object` thành `float`.
- Các giá trị `NaN` trong `horsepower` đã được thay thế bằng `mean` của `horsepower`.
- Tất cả 201 dòng của `horsepower` hiện đã không chứa bắt kỳ `NaN` nào nữa.

## 3. Cột `normalized-losses` => `replace` => `mode`.

In [25]:
# Xem thông tin của `normalized-lossess`
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype 
--------------  ----- 
164 non-null    object
dtypes: object(1)
memory usage: 3.1+ KB


In [26]:
# Show N/A
df[df['normalized-losses'].isna()]['normalized-losses']

,normalized-losses
0,NaN
1,NaN
2,NaN
5,NaN
7,NaN
14,NaN
15,NaN
16,NaN
17,NaN
43,NaN


### Đánh Giá

- Tổng 201 dòng.
- Chỉ có 164 dòng có dữ liệu (non-null).
- Kiểu dữ liệu là `object`.

### Kế Hoạch

- Chuyển dữ liệu sang dạng số để có thể tính `mode`.
- Dùng `replace` để thay thế các giá trị null/NaN với `mode`.
- Kiểm tra lại kết quả.

In [27]:
# Convert data type
df['normalized-losses'] = df['normalized-losses'].astype('float')
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
164 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [28]:
# Calculate mode, get the very first value only
mode_normalized_losses = df['normalized-losses'].mode()[0]
mode_normalized_losses

np.float64(161.0)

In [29]:
# dry-run/preview the replace method
df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=False).info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


In [30]:
# Apply the replace method
df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=True)
df['normalized-losses'].info()

<class 'pandas.core.series.Series'>
Index: 201 entries, 0 to 204
Series name: normalized-losses
Non-Null Count  Dtype  
--------------  -----  
201 non-null    float64
dtypes: float64(1)
memory usage: 3.1 KB


/tmp/ipykernel_23666/1817815612.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['normalized-losses'].replace(to_replace=np.nan, value=mode_normalized_losses, inplace=True)


In [31]:
# Show N/A
df[df['normalized-losses'].isna()]['normalized-losses']

,normalized-losses


In [32]:
# Preview
df['normalized-losses']

,normalized-losses
0,161.0
1,161.0
2,161.0
3,164.0
4,164.0
...,...
200,95.0
201,95.0
202,95.0
203,95.0


### Kết Quả

- `normalized-lossess` đã được thay thế các giá trị null/NaN với mode.
- Hiện tại không còn tồn tại dòng NaN nào trong `normalized-lossess` nữa.

## 4. Bước cuối cùng, `dropna` dòng cho toàn bộ dữ liệu.

### Kế Hoạch

- Xác định kích thước hiện tại của `df`.
- Kiểm tra số lượng hoặc/và xem trước các dòng/cột có NaN trong `df`.
- Thực hiện `dropna`.
- Kiểm tra lại kích thước của `df` và các giá trị NaN nếu còn tồn tại.

In [33]:
# df shape
df.shape

(201, 26)

In [34]:
# các dòng có NaN (ở bất kỳ cột nào)
df[df.isna().any(axis=1)]

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
27,1,148.0,dodge,gas,turbo,NaN,sedan,fwd,front,93.7,...,98,mpfi,3.03,3.39,7.6,102.000000,5500,24,30,8558
55,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,10945
56,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,11845
57,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,70,4bbl,NaN,NaN,9.4,101.000000,6000,17,23,13645
58,3,150.0,mazda,gas,std,two,hatchback,rwd,front,95.3,...,80,mpfi,NaN,NaN,9.4,135.000000,6000,16,23,15645
63,0,161.0,mazda,diesel,std,NaN,sedan,fwd,front,98.8,...,122,idi,3.39,3.39,22.7,64.000000,4650,36,42,10795
130,0,161.0,renault,gas,std,four,wagon,fwd,front,96.1,...,132,mpfi,3.46,3.90,8.7,103.396985,NaN,23,31,9295
131,2,161.0,renault,gas,std,two,hatchback,fwd,front,96.1,...,132,mpfi,3.46,3.90,8.7,103.396985,NaN,23,31,9895


In [35]:
# drop na for all the dataframe: dry-run
df.dropna(axis=0, subset=df.columns, inplace=False).info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [36]:
# drop na for all the dataframe: execute
df.dropna(axis=0, subset=df.columns, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [37]:
# Kiểm tra lại các dòng có NaN (ở bất kỳ cột nào)
df[df.isna().any(axis=1)]

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price


In [38]:
# Kiểm lại shape
df.shape

(193, 26)

### Kết Quả

- `df` hiện không tồn tại giá trị NaN nào nữa.
- Số dòng giảm xuống tương ứng với số dòng đã tồn tại NaN trước đó (201 - 8 = 193).


## 5. Chuyển kiểu phù hợp cho các cột còn lại.

In [39]:
# df info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [40]:
# sample
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,engine-size,fuel-system,bore,stroke,compression-ratio,horsepower,peak-rpm,city-mpg,highway-mpg,price
126,3,161.0,porsche,gas,std,two,hardtop,rwd,rear,89.5,...,194,mpfi,3.74,2.90,9.5,207.0,5900,17,25,32528


### Nhận Xét

Dựa trên `df.info()` và `.sample()` ở tên, chúng ta xác định các cột cần đổi như dưới đây.

- Lý do: Nhận đúng dạng kiểu số nếu đang nhận là object.

Các cột cần đổi:

```
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  

 18  bore               193 non-null    object -> float
 19  stroke             193 non-null    object -> float

 22  peak-rpm           193 non-null    object -> int

 25  price              193 non-null    object -> float
```

In [41]:
# Kiểm lại giá trị thực tế của dữ liệu
df.loc[12, ['bore', 'stroke', 'peak-rpm', 'price']]

,12
bore,3.31
stroke,3.19
peak-rpm,4250
price,20970


In [42]:
# Kiểm tra kiểu dữ liệu trước khi thực hiện
df[['bore', 'stroke', 'peak-rpm', 'price']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   bore      193 non-null    object
 1   stroke    193 non-null    object
 2   peak-rpm  193 non-null    object
 3   price     193 non-null    object
dtypes: object(4)
memory usage: 11.6+ KB


In [43]:
# Đổi kiểu dữ liệu
# Dù price trông rất giống int nhưng sẽ dùng float từ thuộc tính tự nhiên của nó
df['bore'] = df['bore'].astype('float')
df['stroke'] = df['stroke'].astype('float')
df['peak-rpm'] = df['peak-rpm'].astype('int')
df['price'] = df['price'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [44]:
# Kiểm lại kiểu dữ liệu
df[['bore', 'stroke', 'peak-rpm', 'price']].info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   bore      193 non-null    float64
 1   stroke    193 non-null    float64
 2   peak-rpm  193 non-null    int64  
 3   price     193 non-null    float64
dtypes: float64(3), int64(1)
memory usage: 11.6 KB


### Kết Quả

- Các cột đã có kiểu dữ liệu phù hợp với giá trị thực tế.
- Cụ thể các cột đã được cập nhật là: 'bore', 'stroke', 'peak-rpm', 'price'.

## 6. Chọn ra các cột số, chuẩn hóa thành `zscore`.

### Nhận Xét

- `df` có một số lượng lớn các cột có dạng số (numeric), bao gồm `int` và `float`.

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   make               193 non-null    object 
 3   fuel-type          193 non-null    object 
 4   aspiration         193 non-null    object 
 5   num-of-doors       193 non-null    object 
 6   body-style         193 non-null    object 
 7   drive-wheels       193 non-null    object 
 8   engine-location    193 non-null    object 
 9   wheel-base         193 non-null    float64
 10  length             193 non-null    float64
 11  width              193 non-null    float64
 12  height             193 non-null    float64
 13  curb-weight        193 non-null    int64  
 14  engine-type        193 non-null    object 
 15  num-of-cylinders   193 non-null    object 
 16  engine-size        193 non-null

In [46]:
df.dtypes

,0
symboling,int64
normalized-losses,float64
make,object
fuel-type,object
aspiration,object
num-of-doors,object
body-style,object
drive-wheels,object
engine-location,object
wheel-base,float64


### Kế Hoạch

- Lọc ra những cột có `dtype` là kiểu số.
- Sử dụng `sklearn.preprocessing.StandardScaler` để `transform` từng cột tương ứng.

In [47]:
# Chọn ra những cột có dạng số
df.select_dtypes(include=['int64', 'float64']).info()

<class 'pandas.core.frame.DataFrame'>
Index: 193 entries, 0 to 204
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          193 non-null    int64  
 1   normalized-losses  193 non-null    float64
 2   wheel-base         193 non-null    float64
 3   length             193 non-null    float64
 4   width              193 non-null    float64
 5   height             193 non-null    float64
 6   curb-weight        193 non-null    int64  
 7   engine-size        193 non-null    int64  
 8   bore               193 non-null    float64
 9   stroke             193 non-null    float64
 10  compression-ratio  193 non-null    float64
 11  horsepower         193 non-null    float64
 12  peak-rpm           193 non-null    int64  
 13  city-mpg           193 non-null    int64  
 14  highway-mpg        193 non-null    int64  
 15  price              193 non-null    float64
dtypes: float64(10), int64(6)
memory

In [48]:
# Lấy ra danh sách các cột này
columns_list = df.select_dtypes(include=['int64', 'float64']).columns.to_list()
columns_list

['symboling',
 'normalized-losses',
 'wheel-base',
 'length',
 'width',
 'height',
 'curb-weight',
 'engine-size',
 'bore',
 'stroke',
 'compression-ratio',
 'horsepower',
 'peak-rpm',
 'city-mpg',
 'highway-mpg',
 'price']

In [49]:
# import StandardScaler
from sklearn.preprocessing import StandardScaler
ss = StandardScaler()

In [50]:
# Thực hiện transform zscore
for column in columns_list:
    df[column + '_zscore'] = ss.fit_transform(df[[column]])

In [51]:
# Kiểm lại df
df.shape

(193, 42)

In [52]:
# Xem sample
df.sample()

,symboling,normalized-losses,make,fuel-type,aspiration,num-of-doors,body-style,drive-wheels,engine-location,wheel-base,...,curb-weight_zscore,engine-size_zscore,bore_zscore,stroke_zscore,compression-ratio_zscore,horsepower_zscore,peak-rpm_zscore,city-mpg_zscore,highway-mpg_zscore,price_zscore
23,1,118.0,dodge,gas,turbo,two,hatchback,fwd,front,93.7,...,-0.825205,-0.726193,-1.106537,0.448629,-0.641169,-0.039139,0.856208,-0.208189,-0.115832,-0.660382


In [53]:
# Lấy ra các cột zscore
df[[ column + '_zscore' for column in columns_list]]

,symboling_zscore,normalized-losses_zscore,wheel-base_zscore,length_zscore,width_zscore,height_zscore,curb-weight_zscore,engine-size_zscore,bore_zscore,stroke_zscore,compression-ratio_zscore,horsepower_zscore,peak-rpm_zscore,city-mpg_zscore,highway-mpg_zscore,price_zscore
0,1.786850,0.921206,-1.682379,-0.444024,-0.841263,-2.122598,-0.025713,0.045215,0.513027,-1.808186,-0.288273,0.198569,-0.213359,-0.679053,-0.557058,0.026025
1,1.786850,0.921206,-1.682379,-0.444024,-0.841263,-2.122598,-0.025713,0.045215,0.513027,-1.808186,-0.288273,0.198569,-0.213359,-0.679053,-0.557058,0.398480
2,0.163970,0.921206,-0.720911,-0.251195,-0.184679,-0.615412,0.497764,0.575559,-2.394827,0.702918,-0.288273,1.334283,-0.213359,-0.992963,-0.704134,0.398480
3,0.975410,1.005348,0.142781,0.182672,0.143612,0.180047,-0.427362,-0.461021,-0.517605,0.480415,-0.036204,-0.039139,0.856208,-0.208189,-0.115832,0.082420
4,0.975410,1.005348,0.077596,0.182672,0.237410,0.180047,0.499668,0.189854,-0.517605,0.480415,-0.540341,0.304217,0.856208,-1.149918,-1.292436,0.516227
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
200,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,0.743322,0.310387,1.654084,-0.314238,-0.162238,0.277805,0.642295,-0.365144,-0.409983,0.441241
201,-1.458911,-0.929925,1.658315,1.162889,1.362982,0.682442,0.927967,0.310387,1.654084,-0.314238,-0.363893,1.492755,0.428381,-0.992963,-0.851209,0.713920
202,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,0.857535,1.081795,0.917918,-1.204249,-0.338686,0.806044,0.856208,-1.149918,-1.145360,1.016345
203,-1.458911,-0.929925,1.658315,1.162889,1.409880,0.682442,1.247763,0.406813,-1.180154,0.480415,3.240689,0.066509,-0.641186,0.105721,-0.557058,1.138431


### Kết Quả

- Có tổng số 16 cột dạng số.
- Tất cả các cột dạng số đã được tạo cột zscore, tương ứng 16 cột thêm mới.
- Kích thước của `df` hiện tại là (193, 42), với 42 = 26 cột đã có + 16 cột mới.